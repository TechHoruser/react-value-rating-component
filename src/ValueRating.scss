.valueRating {
  --good-color: #1a92b6;
  --bad-color: #c04324;
  --neutral-color: #5c4f4d;
  --icon-height: 1rem;
  --icon-deep: 0.2rem;
  --icon-width: calc(var(--icon-height) / 0.75);

  $transactionTime: 300ms;

  width: var(--icon-width);
  height: var(--icon-height);

  .icon {
    width: 100%;
    height: 100%;
    transform-style: preserve-3d;

    .triangle {
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      transform-style: preserve-3d;

      span {
        @mixin triangle {
          position: absolute;
          top: 0;
          left: 0;
          width: 100%;
          height: 100%;
          transform-origin: bottom;
          clip-path: polygon(50% 0%, 0% 100%, 100% 100%);
        }

        &.front {
          transform: translateZ(calc(var(--icon-deep) * 0.5));

          @include triangle;
        }

        &.back {
          transform: translateZ(calc(var(--icon-deep) * -0.5));

          @include triangle;
        }

        &.floor {
          position: absolute;
          bottom: calc(var(--icon-deep) * -1);
          left: 0;
          width: 100%;
          height: var(--icon-deep);
          transform: rotateX(90deg) translateY(calc(var(--icon-deep) * 0.5 * -1));
          transform-origin: top;
        }

        transition: background-color $transactionTime ease;
      }

      &.good {
        transition: transform $transactionTime ease;

        span {
          background-color: var(--good-color);
        }
      }

      &.bad {
        transform: rotateX(180deg);
        transition: transform $transactionTime ease;

        span {
          background-color: var(--bad-color);
        }
      }

      &.neutral {
        transition: transform calc($transactionTime) ease;

        &.shape1 {
          transform: rotateX(90deg) translateZ(calc(var(--icon-height) * 0.15));
        }

        &.shape2 {
          transform: rotateX(90deg) translateZ(calc(var(--icon-height) * 0.15 * -1));
        }

        span {
          background-color: var(--neutral-color);
        }
      }

      &.mid-neutral {
        transition: transform calc($transactionTime) ease;

        &.shape1 {
          transform: rotateX(90deg);
        }

        &.shape2 {
          transform: rotateX(90deg);
        }

        span {
          background-color: var(--neutral-color);
        }
      }

      @keyframes rotate {
        0% {
          transform: rotateX(0);
        }

        50% {
          transform: rotateX(180deg);
        }

        100% {
          transform: rotateX(0deg);
        }
      }
    }
  }
}
