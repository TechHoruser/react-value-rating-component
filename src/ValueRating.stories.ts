import type { Meta, StoryObj } from "@storybook/react";

import { ValueRating } from "./ValueRating";

const meta = {
	title: "ValueRating",
	component: ValueRating,
	argTypes: {
		value: { control: { type: 'number' } },
		endBad: { control: { type: 'number' } },
		startGood: { control: { type: 'number' } },
	},
} satisfies Meta<typeof ValueRating>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
	args: {
		value: 0,
		endBad: 0.5,
		startGood: 0.5,
	},
};
