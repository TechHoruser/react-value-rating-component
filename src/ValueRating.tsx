import { useEffect, useState } from "react";
import "./ValueRating.scss";

type StatusOption = 'bad' | 'good' | 'neutral' | 'mid-neutral';

interface ValueRatingProps {
	value: number,
	endBad?: number,
	startGood?: number,
}

const getNewStatus = (value: number, endBad: number, startGood: number) => {
	const newStatus: StatusOption = (value < endBad
		? 'bad'
		: (value > startGood
			? 'good'
			: 'neutral')
	);

	return newStatus;
}

export function ValueRating({
	value,
	endBad = 0,
	startGood = 0,
}: ValueRatingProps) {
	const [status, setStatus] = useState<StatusOption>(getNewStatus(value, endBad, startGood))

	useEffect(
		() => {
			const changeStatusWithCalcs = () => {
				let newStatus: StatusOption = getNewStatus(value, endBad, startGood);

				if (status === 'mid-neutral') {
					setTimeout(() => setStatus(newStatus), 300);
					return;
				}

				if (
					(status === 'neutral' && ['good', 'bad'].includes(newStatus))
					|| (newStatus === 'neutral' && ['good', 'bad'].includes(status))
				) {
					newStatus = 'mid-neutral';
				}

				setStatus(newStatus);
			}

			changeStatusWithCalcs();
		},
		[value, endBad, startGood, status]
	)

	return (
		<div className="valueRating">
			<div className="icon">
				<div className={`triangle shape1 ${status}`}>
					<span className="front"></span>
					<span className="back"></span>
					<span className="floor"></span>
				</div>
				<div className={`triangle shape2 ${status}`}>
					<span className="front"></span>
					<span className="back"></span>
					<span className="floor"></span>
				</div>
			</div>
		</div>
	);
}
