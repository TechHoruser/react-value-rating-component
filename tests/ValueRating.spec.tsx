import { render } from '@testing-library/react';
import { ValueRating } from '../src';

describe('ValueRating component', () => {
	it('renders with default props', () => {
		const { container } = render(<ValueRating value={50} />);
		expect(container).toMatchSnapshot();
	});

	it('renders with custom props', () => {
		const { container } = render(<ValueRating value={30} endBad={20} startGood={70} />);
		expect(container).toMatchSnapshot();
	});

	it('changes status when value changes', () => {
		const { container, rerender } = render(<ValueRating value={30} endBad={20} startGood={70} />);

		// Initial snapshot
		expect(container).toMatchSnapshot();

		// Update the value prop
		rerender(<ValueRating value={80} endBad={20} startGood={70} />);

		// Updated snapshot
		expect(container).toMatchSnapshot();
	});

	it('changes status to "mid-neutral" and then updates to new status', async () => {
		const { container, rerender } = render(<ValueRating value={30} endBad={20} startGood={70} />);

		// Initial snapshot
		expect(container).toMatchSnapshot();

		// Update the value to trigger "mid-neutral"
		jest.useFakeTimers();
		rerender(<ValueRating value={45} endBad={20} startGood={70} />);
		jest.runAllTimers();
		jest.useRealTimers();

		// Snapshot after "mid-neutral"
		expect(container).toMatchSnapshot();

		// Update the value to new status
		rerender(<ValueRating value={90} endBad={20} startGood={70} />);

		// Updated snapshot
		expect(container).toMatchSnapshot();
	});
});