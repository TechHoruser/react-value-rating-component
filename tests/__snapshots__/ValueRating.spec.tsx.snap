// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueRating component changes status to "mid-neutral" and then updates to new status 1`] = `
<div>
  <div
    class="valueRating"
  >
    <div
      class="icon"
    >
      <div
        class="triangle shape1 neutral"
      >
        <span
          class="front"
        />
        <span
          class="back"
        />
        <span
          class="floor"
        />
      </div>
      <div
        class="triangle shape2 neutral"
      >
        <span
          class="front"
        />
        <span
          class="back"
        />
        <span
          class="floor"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ValueRating component changes status to "mid-neutral" and then updates to new status 2`] = `
<div>
  <div
    class="valueRating"
  >
    <div
      class="icon"
    >
      <div
        class="triangle shape1 neutral"
      >
        <span
          class="front"
        />
        <span
          class="back"
        />
        <span
          class="floor"
        />
      </div>
      <div
        class="triangle shape2 neutral"
      >
        <span
          class="front"
        />
        <span
          class="back"
        />
        <span
          class="floor"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ValueRating component changes status to "mid-neutral" and then updates to new status 3`] = `
<div>
  <div
    class="valueRating"
  >
    <div
      class="icon"
    >
      <div
        class="triangle shape1 mid-neutral"
      >
        <span
          class="front"
        />
        <span
          class="back"
        />
        <span
          class="floor"
        />
      </div>
      <div
        class="triangle shape2 mid-neutral"
      >
        <span
          class="front"
        />
        <span
          class="back"
        />
        <span
          class="floor"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ValueRating component changes status when value changes 1`] = `
<div>
  <div
    class="valueRating"
  >
    <div
      class="icon"
    >
      <div
        class="triangle shape1 neutral"
      >
        <span
          class="front"
        />
        <span
          class="back"
        />
        <span
          class="floor"
        />
      </div>
      <div
        class="triangle shape2 neutral"
      >
        <span
          class="front"
        />
        <span
          class="back"
        />
        <span
          class="floor"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ValueRating component changes status when value changes 2`] = `
<div>
  <div
    class="valueRating"
  >
    <div
      class="icon"
    >
      <div
        class="triangle shape1 mid-neutral"
      >
        <span
          class="front"
        />
        <span
          class="back"
        />
        <span
          class="floor"
        />
      </div>
      <div
        class="triangle shape2 mid-neutral"
      >
        <span
          class="front"
        />
        <span
          class="back"
        />
        <span
          class="floor"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ValueRating component renders with custom props 1`] = `
<div>
  <div
    class="valueRating"
  >
    <div
      class="icon"
    >
      <div
        class="triangle shape1 neutral"
      >
        <span
          class="front"
        />
        <span
          class="back"
        />
        <span
          class="floor"
        />
      </div>
      <div
        class="triangle shape2 neutral"
      >
        <span
          class="front"
        />
        <span
          class="back"
        />
        <span
          class="floor"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ValueRating component renders with default props 1`] = `
<div>
  <div
    class="valueRating"
  >
    <div
      class="icon"
    >
      <div
        class="triangle shape1 good"
      >
        <span
          class="front"
        />
        <span
          class="back"
        />
        <span
          class="floor"
        />
      </div>
      <div
        class="triangle shape2 good"
      >
        <span
          class="front"
        />
        <span
          class="back"
        />
        <span
          class="floor"
        />
      </div>
    </div>
  </div>
</div>
`;
